### PR TITLE
chore(dgraph): Expose cache metrics to prometheus

### PR DESCRIPTION
**Pull request by ahsanbarkati**
Fixes DGRAPH-2300

This PR exposes the hit-ratio of several caches in dgraph to prometheus. The exposed caches includes the cache of p store, WAL store, zero and the posting list.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6426)

 
 

Docs Preview: [Dgraph Preview](https://dgraph-a9adbac622-94963.surge.sh)

### CODE REVIEW

**Review by jarifibrahim — APPROVED** [review]
:lgtm: please get review from @danielmai. @ahsanbarkati  I have not verified running the code locally so please do verify that the metrics show up correctly on grafana. 

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-:-MGrn8n6BFsMvJOXd1vu:biez49q)* status: 0 of 5 files reviewed, 9 unresolved discussions (waiting on @ahsanbarkati, @jarifibrahim, @manishrjain, @martinmr, and @vvbalaji-dgraph)

___
*[x/metrics.go, line 84 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MGrm_1K4WIclNWDy97T:-MGrm_1K4WIclNWDy97U:bgxdl6h) ([raw file](https://github.com/dgraph-io/dgraph/blob/4827e13be06c251fa80f7efc7fd82e775a4b785a/x/metrics.go#L84)):*
> ```Go
> 	MaxAssignedTs = stats.Int64("max_assigned_ts",
> 		"Latest max assigned timestamp", stats.UnitDimensionless)
> 	// TxnAborts records count of aborted transactions
> ```

nit: missing period.
___
*[x/metrics.go, line 90 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MGrmg6kAz7yzISAHTC6:-MGrmg6kAz7yzISAHTC7:by6cwsu) ([raw file](https://github.com/dgraph-io/dgraph/blob/4827e13be06c251fa80f7efc7fd82e775a4b785a/x/metrics.go#L90)):*
> ```Go
> 	PBlockHitRatio = stats.Float64("hit_ratio_p_block",
> 		"Hit ratio of p store block cache", stats.UnitDimensionless)
> 	// PIndexHitRatio records the hit ratio of p store block cache
> ```

p store index cache.
___
*[x/metrics.go, line 93 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MGrmb6TCLMzgtjmip2S:-MGrmb6TCLMzgtjmip2T:bd4a4w6) ([raw file](https://github.com/dgraph-io/dgraph/blob/4827e13be06c251fa80f7efc7fd82e775a4b785a/x/metrics.go#L93)):*
> ```Go
> 	PIndexHitRatio = stats.Float64("hit_ratio_p_index",
> 		"Hit ratio of p store index cache", stats.UnitDimensionless)
> 	// WBlockHitRatio records the hit ratio of p store block cache
> ```

nit: missing period.

records the hit ratio of w store block cache
___
*[x/metrics.go, line 96 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MGrmnPoCbh8LjSS5zbI:-MGrmnPoCbh8LjSS5zbJ:b-f9tw9) ([raw file](https://github.com/dgraph-io/dgraph/blob/4827e13be06c251fa80f7efc7fd82e775a4b785a/x/metrics.go#L96)):*
> ```Go
> 	WBlockHitRatio = stats.Float64("hit_ratio_w_block",
> 		"Hit ratio of w store block cache", stats.UnitDimensionless)
> 	// WIndexHitRatio records the hit ratio of p store block cache
> ```

w store index cache
___
*[x/metrics.go, line 102 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MGrmxFcAButG-nj41Wi:-MGrmxFcAButG-nj41Wj:b-lvx0fl) ([raw file](https://github.com/dgraph-io/dgraph/blob/4827e13be06c251fa80f7efc7fd82e775a4b785a/x/metrics.go#L102)):*
> ```Go
> 	ZWBlockHitRatio = stats.Float64("hit_ratio_zw_block",
> 		"Hit ratio of w store block cache", stats.UnitDimensionless)
> 	// ZWIndexHitRatio records the hit ratio of zw store block cache
> ```

index cache
___
*[x/x.go, line 1056 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MGmlD9VAVVGqhS4zMxi:-MGmlD9VAVVGqhS4zMxj:bb618wr) ([raw file](https://github.com/dgraph-io/dgraph/blob/69f61fe2b1084fb07b31a1d59306a91852c0879e/x/x.go#L1056)):*
> ```Go
> 		glog.V(2).Infof("%s metrics %+v %+v", prefix, metrics, le)
> 
> 		// If the mean life expectancy is leff than 5 seconds, the cache
> ```

less than 10 seconds
___
*[x/x.go, line 1050 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MGrmE3R-uTSADiQqPSY:-MGrmE3R-uTSADiQqPSZ:b-j13zwd) ([raw file](https://github.com/dgraph-io/dgraph/blob/4827e13be06c251fa80f7efc7fd82e775a4b785a/x/x.go#L1050)):*
> ```Go
> 			metrics = db.IndexCacheMetrics()
> 			ostats.Record(context.Background(), ZWIndexHitRatio.M(metrics.Ratio()))
> 		default:
> ```

Posting list cache is missing here
___
*[x/x.go, line 1061 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MGrlyjE6DvRFhsQf-OV:-MGrlyjE6DvRFhsQf-OW:b-csxmt0) ([raw file](https://github.com/dgraph-io/dgraph/blob/4827e13be06c251fa80f7efc7fd82e775a4b785a/x/x.go#L1061)):*
> ```Go
> 
> 		// If the mean life expectancy is leff than 5 seconds, the cache
> 		// might be under contention.
> ```

might be too small
___
*[x/x.go, line 1063 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MGrm1IR-pUETBnVFDca:-MGrm1IR-pUETBnVFDcb:b-wk8zgx) ([raw file](https://github.com/dgraph-io/dgraph/blob/4827e13be06c251fa80f7efc7fd82e775a4b785a/x/x.go#L1063)):*
> ```Go
> 		// might be under contention.
> 		lifeTooShort := le.Count > 0 && float64(le.Sum)/float64(le.Count) < 5
> 		// Check misses and hits to ensure we're not looking at a closed cache.
> ```

out of date comment.

**Comment by ahsanbarkati** [comment]
___
*[x/x.go, line 1050 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MGrmE3R-uTSADiQqPSY:-MHBd_zS9onReo5zX0tS:b-mdd989) ([raw file](https://github.com/dgraph-io/dgraph/blob/4827e13be06c251fa80f7efc7fd82e775a4b785a/x/x.go#L1050)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Posting list cache is missing here
</blockquote></details>

It is recordedhere: https://github.com/dgraph-io/dgraph/blob/4827e13be06c251fa80f7efc7fd82e775a4b785a/posting/lists.go#L168

**Review by ahsanbarkati — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6426)* status: 0 of 5 files reviewed, 7 unresolved discussions (waiting on @ahsanbarkati, @jarifibrahim, @manishrjain, @martinmr, and @vvbalaji-dgraph)

___
*[x/metrics.go, line 90 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MGrmg6kAz7yzISAHTC6:-MHBci602CWJNNWZbfC1:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/4827e13be06c251fa80f7efc7fd82e775a4b785a/x/metrics.go#L90)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

p store index cache.
</blockquote></details>

Done.
___
*[x/metrics.go, line 93 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MGrmb6TCLMzgtjmip2S:-MHBcimtEqR339wUosqP:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/4827e13be06c251fa80f7efc7fd82e775a4b785a/x/metrics.go#L93)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

nit: missing period.

records the hit ratio of w store block cache
</blockquote></details>

Done.
___
*[x/x.go, line 1056 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MGmlD9VAVVGqhS4zMxi:-MHBe6rA1FrwSSp3SikV:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/69f61fe2b1084fb07b31a1d59306a91852c0879e/x/x.go#L1056)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

less than 10 seconds
</blockquote></details>

Done.
___
*[x/x.go, line 1061 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MGrlyjE6DvRFhsQf-OV:-MHBeETs0FY74gK_XGxF:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/4827e13be06c251fa80f7efc7fd82e775a4b785a/x/x.go#L1061)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

might be too small
</blockquote></details>

Done.
___
*[x/x.go, line 1063 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MGrm1IR-pUETBnVFDca:-MHBeFa0DOqo42gVXL8v:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/4827e13be06c251fa80f7efc7fd82e775a4b785a/x/x.go#L1063)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

out of date comment.
</blockquote></details>

Done.

**Review by danielmai — CHANGES_REQUESTED** [review]
> Reviewed 1 of 5 files at r1, 2 of 3 files at r2, 1 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6426)* status: 4 of 5 files reviewed, 11 unresolved discussions (waiting on @ahsanbarkati, @jarifibrahim, @manishrjain, @martinmr, and @vvbalaji-dgraph)

___
*[posting/lists.go, line 166 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MHMi1Q37A8XE0rhKN29:-MHMi1Q37A8XE0rhKN2A:bnn97h4) ([raw file](https://github.com/dgraph-io/dgraph/blob/6555b27f2f17278e8df8ab3f4da8a013aa8e8447/posting/lists.go#L166)):*
> ```
> glog.V(2).Infof("Posting list cache metrics: %s", m)
> ```
Do we still need this glog? The ticker is very frequent now (every 10 seconds instead of 5 minutes). The metrics are recorded via ostats anyway.
___
*[x/metrics.go, line 91 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MHMichABcxcg8Gt6cNW:-MHMichABcxcg8Gt6cNX:bg18ows) ([raw file](https://github.com/dgraph-io/dgraph/blob/6555b27f2f17278e8df8ab3f4da8a013aa8e8447/x/metrics.go#L91)):*
> ```
> PIndexHitRatio = stats.Float64("hit_ratio_p_index",
> ```
Let's use the full names for these metrics. e.g., instead of p, say postings, and instead of w, say wal.

w and zw can use the same metric names. They're different processes anyway, so they wouldn't collide. e.g., `hit_ratio_wal_block` for both Alpha (w) and Zero (zw) write-ahead logs.

So, I'm in favor of having the following be the metric names (the dgraph_ prefix is already added):

* dgraph_hit_ratio_postings_block
* dgraph_hit_ratio_postings_index
* dgraph_hit_ratio_wal_block (Alpha and Zero)
* dgraph_hit_ratio_wal_index (Alpha and Zero)
* dgraph_hit_ratio_posting_cache for the posting list cache hit ratio.
___
*[x/x.go, line 1058 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MHMlDwoCWmgA_EGMW8e:-MHMlDwp6bhWG18ltFm3:bm1ngt5) ([raw file](https://github.com/dgraph-io/dgraph/blob/6555b27f2f17278e8df8ab3f4da8a013aa8e8447/x/x.go#L1058)):*
> ```
> glog.V(2).Infof("%s metrics %+v %+v", prefix, metrics, le)
> ```
How frequently is this logged?
___
*[x/x.go, line 1064 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MHMl1VHEhE58I83UvPd:-MHMl1VHEhE58I83UvPe:b8p25dq) ([raw file](https://github.com/dgraph-io/dgraph/blob/6555b27f2f17278e8df8ab3f4da8a013aa8e8447/x/x.go#L1064)):*
> ```
> bool(glog.V(2))
> ```
Does this really need `bool()`?

**Review by ahsanbarkati — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6426)* status: 4 of 5 files reviewed, 11 unresolved discussions (waiting on @ahsanbarkati, @danielmai, @jarifibrahim, @manishrjain, @martinmr, and @vvbalaji-dgraph)

___
*[posting/lists.go, line 166 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MHMi1Q37A8XE0rhKN29:-MHQI8Jt1hmpV1B-pL6E:bo8v4yq) ([raw file](https://github.com/dgraph-io/dgraph/blob/6555b27f2f17278e8df8ab3f4da8a013aa8e8447/posting/lists.go#L166)):*
<details><summary><i>Previously, danielmai (Daniel Mai) wrote…</i></summary><blockquote>

> ```
> glog.V(2).Infof("Posting list cache metrics: %s", m)
> ```
Do we still need this glog? The ticker is very frequent now (every 10 seconds instead of 5 minutes). The metrics are recorded via ostats anyway.
</blockquote></details>

Removed glog
___
*[x/metrics.go, line 91 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MHMichABcxcg8Gt6cNW:-MHQZ2We-z4SkNal7GLe:b-og0q5w) ([raw file](https://github.com/dgraph-io/dgraph/blob/6555b27f2f17278e8df8ab3f4da8a013aa8e8447/x/metrics.go#L91)):*
<details><summary><i>Previously, danielmai (Daniel Mai) wrote…</i></summary><blockquote>

> ```
> PIndexHitRatio = stats.Float64("hit_ratio_p_index",
> ```
Let's use the full names for these metrics. e.g., instead of p, say postings, and instead of w, say wal.

w and zw can use the same metric names. They're different processes anyway, so they wouldn't collide. e.g., `hit_ratio_wal_block` for both Alpha (w) and Zero (zw) write-ahead logs.

So, I'm in favor of having the following be the metric names (the dgraph_ prefix is already added):

* dgraph_hit_ratio_postings_block
* dgraph_hit_ratio_postings_index
* dgraph_hit_ratio_wal_block (Alpha and Zero)
* dgraph_hit_ratio_wal_index (Alpha and Zero)
* dgraph_hit_ratio_posting_cache for the posting list cache hit ratio.
</blockquote></details>

I have changed the names to full names. I am not sure if we can have same name for w and zw. If I do this, I get an error, complaining that  "a different view with the same name is already registered".
___
*[x/x.go, line 1058 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MHMlDwoCWmgA_EGMW8e:-MHQK7sgAaLyGxIW4vHi:b-lx6mrn) ([raw file](https://github.com/dgraph-io/dgraph/blob/6555b27f2f17278e8df8ab3f4da8a013aa8e8447/x/x.go#L1058)):*
<details><summary><i>Previously, danielmai (Daniel Mai) wrote…</i></summary><blockquote>

> ```
> glog.V(2).Infof("%s metrics %+v %+v", prefix, metrics, le)
> ```
How frequently is this logged?
</blockquote></details>

This is logged every 10 second. 
___
*[x/x.go, line 1064 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-MHMl1VHEhE58I83UvPd:-MHQKgcV1zILX-QjAWbN:bot1rtv) ([raw file](https://github.com/dgraph-io/dgraph/blob/6555b27f2f17278e8df8ab3f4da8a013aa8e8447/x/x.go#L1064)):*
<details><summary><i>Previously, danielmai (Daniel Mai) wrote…</i></summary><blockquote>

> ```
> bool(glog.V(2))
> ```
Does this really need `bool()`?
</blockquote></details>

It is required because `(lifeTooShort || hitRatioTooLow)` is bool and `glog.V(2)` is of type `Verbose`. So there is a mismatch of type while doing an `&&`.

**Comment by jarifibrahim** [comment]
I ran this PR with an existing cluster and I have two comments
1. Reduce the histogram logs. When running with v2, the logs were filled up with histogram output (maybe because cache had very low hit ratio)
2. The metric should zero even when the caches are disabled, @ahsanbarkati please fix this.

**Comment by ahsanbarkati** [comment]
> I ran this PR with an existing cluster and I have two comments
> 
>     1. Reduce the histogram logs. When running with v2, the logs were filled up with histogram output (maybe because cache had very low hit ratio)
> 
>     2. The metric should zero even when the caches are disabled, @ahsanbarkati please fix this.

Done

File: x/x.go [hunk 1]
```diff
@@ -1019,6 +1021,85 @@ func IsGuardian(groups []string) bool {
...
+		switch ct {
+		case "pstore-block":
+			metrics = db.BlockCacheMetrics()
+		case "pstore-index":
+			metrics = db.IndexCacheMetrics()
+		case "WALstore-block":
+			metrics = db.BlockCacheMetrics()
+		case "WALstore-index":
+			metrics = db.IndexCacheMetrics()
+		default:
```
Comment (jarifibrahim) [inline x/x.go:1040]: You don't need this. `metrics` is nil by default

File: x/x.go [hunk 2]
```diff
@@ -1019,6 +1021,85 @@ func IsGuardian(groups []string) bool {
...
+			prefix := fmt.Sprintf("%s-%s", prefix, ct)
+			le := metrics.LifeExpectancySeconds()
+			glog.V(2).Infof("%s metrics %+v %+v", prefix, metrics, le)
+		}
+	}
+
+	ticker := time.NewTicker(period)
+	tickerLog := time.NewTicker(5 * time.Minute)
+	defer ticker.Stop()
+	defer ticker.Stop()
```
Comment (jarifibrahim) [inline x/x.go:1087]: Should be tickerLog. Try to keep the defer statement near the initialization.

**Review by jarifibrahim — APPROVED** [review]

File: posting/lists.go [hunk 3]
```diff
@@ -163,9 +161,10 @@ func Init(ps *badger.DB, cacheSize int64) {
 	x.Check(err)
 	go func() {
 		m := lCache.Metrics
-		ticker := time.NewTicker(5 * time.Minute)
+		ticker := time.NewTicker(10 * time.Second)
```
Comment (jarifibrahim) [inline posting/lists.go:164]: defer ticker.stop()

Comment (jarifibrahim) [reply posting/lists.go:164]: This is only recording the metrics so 10 seconds is okay.

File: x/x.go (hunk 2 above)
Comment (ahsanbarkati) [reply x/x.go:1083]: Fixed. Thanks.

**Review by danielmai — COMMENTED** [review]
How long does it typically take for metrics to show up? I've been running an instance for over ten minutes and don't see any metrics yet.

```
Dgraph version   : v2.0.0-rc1-699-g0e4b09019
Dgraph codename  : unnamed
Dgraph SHA-256   : 896c143cf48364138e03bc34f471fb799fb8e0846f8e908e7d918955bc0ca762
Commit SHA-1     : 0e4b09019
Commit timestamp : 2020-09-18 19:32:22 +0530
Branch           : ahsan/cache-metrics
Go version       : go1.14.4
```

Cluster's been up for 20 minutes and I ran live loader with 1-million movie data set:
```
$ curl -s localhost:8080/health | jq '.[0].uptime'
1230
```

I don't see any of the PR's hit ratio metrics:
```
$ curl -s localhost:8080/debug/prometheus_metrics | grep hit
# HELP badger_v2_lsm_bloom_hits_total Total number of LSM bloom hits
# TYPE badger_v2_lsm_bloom_hits_total untyped
badger_v2_lsm_bloom_hits_total{level="l0"} 2.127442e+06
```

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-:-MHl-Erk-X-YKGRXB8Ko:b-mlhupl)* status: 1 of 5 files reviewed, 13 unresolved discussions (waiting on @ahsanbarkati, @danielmai, @jarifibrahim, @manishrjain, @martinmr, and @vvbalaji-dgraph)

**Comment by jarifibrahim** [comment]
@danielmai you need to enable the cache. It is disabled by default.

**Review by danielmai — COMMENTED** [review]
Ah, got it. Setting `--cache_mb` did the trick. Now I see metrics:

```
$ curl -s localhost:8080/debug/prometheus_metrics | grep ^dgraph_hit
dgraph_hit_ratio_postings_block{method="",status=""} 0.4626865671641791
dgraph_hit_ratio_postings_index{method="",status=""} 0.9424460431654677
dgraph_hit_ratio_wal_block{method="",status=""} 0.05804464973056197
dgraph_hit_ratio_wal_index{method="",status=""} 0.99984040006384
```

Looks like this PR is :lgtm: from a metrics perspective. On the other hand, now that I see these metrics, should I expect wal_block cache hit rate to be this bad? I set ` --cache_mb=2048 --cache_percentage="0,25,25,40,10"` (40% block cache for wal) and I see 5% hit rate. And, these "Cache might be too small WALstore-block" logs spitting out all the time:

```
W0921 17:27:22.583839      16 x.go:1073] ======== Cache might be too small WALstore-block =====
W0921 17:27:22.583862      16 x.go:1074] Metric: hit: 757 miss: 12288 keys-added: 2591 keys-updated: 39 keys-evicted: 2564 cost-added: 106812668004 cost-evicted: 105996684963 sets-dropped: 0 sets-rejected: 9655 gets-dropped: 6208 gets-kept: 6592 gets-total: 13045 hit-ratio: 0.06
W0921 17:27:22.583898      16 x.go:1075] Life expectancy: 
 -- Histogram: 
Min value: 0 
Max value: 59 
Mean: 0.58 
[0 B, 2 B) 2538 98.99% 
[4 B, 8 B) 1 0.04% 
[32 B, 64 B) 25 0.98% 
 --
W0921 17:27:32.583205      16 x.go:1073] ======== Cache might be too small WALstore-block =====
W0921 17:27:32.583228      16 x.go:1074] Metric: hit: 757 miss: 12288 keys-added: 2591 keys-updated: 39 keys-evicted: 2564 cost-added: 106812668004 cost-evicted: 105996684963 sets-dropped: 0 sets-rejected: 9655 gets-dropped: 6208 gets-kept: 6592 gets-total: 13045 hit-ratio: 0.06
W0921 17:27:32.583270      16 x.go:1075] Life expectancy: 
 -- Histogram: 
Min value: 0 
Max value: 59 
Mean: 0.58 
[0 B, 2 B) 2538 98.99% 
[4 B, 8 B) 1 0.04% 
[32 B, 64 B) 25 0.98% 
 --
```

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-:-MHlge2xCTqf-B8iA2VZ:b328e0o)* status: 1 of 5 files reviewed, 13 unresolved discussions (waiting on @ahsanbarkati, @danielmai, @jarifibrahim, @manishrjain, @martinmr, and @vvbalaji-dgraph)

**Review by danielmai — COMMENTED** [review]
I ran into an issue with reporting for new clusters. I get the following logs that show that the w cache hit ratios are 0:
```
W0921 17:56:01.233260      16 x.go:1073] ======== Cache might be too small WALstore-block =====
W0921 17:56:01.233308      16 x.go:1074] Metric: hit: 0 miss: 0 keys-added: 0 keys-updated: 0 keys-evicted: 0 cost-added: 0 cost-evicted: 0 sets-dropped: 0 sets-rejected: 0 gets-dropped: 0 gets-kept: 0 gets-total: 0 hit-ratio: 0.00
W0921 17:56:01.233382      16 x.go:1075] Life expectancy: 
 -- Histogram: 
Min value: 9223372036854775807 
Max value: 0 
Mean: 0.00 
 --
W0921 17:56:01.233427      16 x.go:1073] ======== Cache might be too small WALstore-index =====
W0921 17:56:01.233440      16 x.go:1074] Metric: hit: 0 miss: 0 keys-added: 0 keys-updated: 0 keys-evicted: 0 cost-added: 0 cost-evicted: 0 sets-dropped: 0 sets-rejected: 0 gets-dropped: 0 gets-kept: 0 gets-total: 0 hit-ratio: 0.00
W0921 17:56:01.233480      16 x.go:1075] Life expectancy: 
 -- Histogram: 
Min value: 9223372036854775807 
Max value: 0 
Mean: 0.00 
 --
```

```
dgraph_hit_ratio_postings_block{method="",status=""} 0.9090023201856149
dgraph_hit_ratio_postings_index{method="",status=""} 0.9999715066919383
dgraph_hit_ratio_wal_block{method="",status=""} 0
dgraph_hit_ratio_wal_index{method="",status=""} 0
```

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-:-MHloVPE4ZYnMp6wlYhU:bwjv832)* status: 1 of 5 files reviewed, 13 unresolved discussions (waiting on @ahsanbarkati, @danielmai, @jarifibrahim, @manishrjain, @martinmr, and @vvbalaji-dgraph)

**Comment by jarifibrahim** [comment]
@danielmai Can you show your config as well? If you don't have a cache enabled. the metric shows zero (I thought Ahsan fixed this).

**Comment by ahsanbarkati** [comment]
@jarifibrahim It has been fixed, we don't record 0 if the metrics is nil. There is some other issue. Having the exact configuration will help.

**Comment by danielmai** [comment]
This is the config I'm running:

```
version: "3.5"
services:
  alpha1:
    image: dgraph/dgraph:latest
    container_name: alpha1
    working_dir: /data/alpha1
    labels:
      cluster: test
    ports:
    - 8080:8080
    - 9080:9080
    volumes:
    - type: bind
      source: $GOPATH/bin
      target: /gobin
      read_only: true
    - type: bind
      source: ./data
      target: /data
      read_only: false
    - type: bind
      source: ./enc_key_file
      target: /secret/enc_key
      read_only: true
    command: /gobin/dgraph alpha -o 0 --my=alpha1:7080
      --lru_mb=1024 --zero=zero1:5080 --logtostderr -v=2 --idx=1 --whitelist=10.0.0.0/8,172.16.0.0/12,192.168.0.0/16
      --encryption_key_file=/secret/enc_key --cache_mb=2048 --cache_percentage="0,25,25,40,10"
  zero1:
    image: dgraph/dgraph:latest
    container_name: zero1
    working_dir: /data/zero1
    labels:
      cluster: test
    ports:
    - 5080:5080
    - 6080:6080
    volumes:
    - type: bind
      source: $GOPATH/bin
      target: /gobin
      read_only: true
    - type: bind
      source: ./data
      target: /data
      read_only: false
    command: /gobin/dgraph zero -o 0 --idx=1
      --my=zero1:5080 --logtostderr -v=2 --bindall
volumes: {}
```

If I start from a fresh cluster, then it starts off as Zero, and then if I restart it I see the metrics populated:

```
$ docker-compose up
```

```
$ curl --silent localhost:8080/debug/prometheus_metrics | grep ^dgraph_hit
dgraph_hit_ratio_postings_block{method="",status=""} 0
dgraph_hit_ratio_postings_index{method="",status=""} 0.5555555555555556
dgraph_hit_ratio_wal_block{method="",status=""} 0
dgraph_hit_ratio_wal_index{method="",status=""} 0
```

Then, if I restart the cluster, then the metrics are updated:

```
$ docker-compose restart
...
$ curl --silent localhost:8080/debug/prometheus_metrics | grep ^dgraph_hit
dgraph_hit_ratio_postings_block{method="",status=""} 0.75
dgraph_hit_ratio_postings_index{method="",status=""} 0.8333333333333334
dgraph_hit_ratio_wal_block{method="",status=""} 0.9310344827586207
dgraph_hit_ratio_wal_index{method="",status=""} 0.9948979591836735
```

**Comment by ahsanbarkati** [comment]
@danielmai The metrics are showing the true values of the hit ratios. They are actually zero. If we load some data by live loader the hit ratios changes.

**Review by danielmai — APPROVED** [review]
:lgtm:

> Reviewed 2 of 3 files at r5, 1 of 2 files at r6, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6426#-:-MIJkPFgBYZi6MUgJ9E9:bnfp4nl)* status: all files reviewed, 13 unresolved discussions (waiting on @ahsanbarkati, @danielmai, @jarifibrahim, @manishrjain, @martinmr, and @vvbalaji-dgraph)

**Pull request merged**